### ISSUE TITLE
reader_concurrency_semaphore: CPU concurrency of 1 is too strict in some cases

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#19017]
The reader concurrency semaphore restricts the concurrency of reads that require CPU (intention: they read from the cache) to 1, meaning that if there is even a single active read which declares that it needs just CPU to proceed, no new read is admitted. This is meant to keep the concurrency of reads in the cache at 1. The idea is that concurrency in the cache is not useful: it just leads to the reactor rotating between these reads, all of the finishing later then they could if they were the only active read in the cache.
This was observed to backfire in the case where there reads from a single table are mostly very fast, but on some keys are very slow (hint: collection full of tombstones). In this case the slow read keeps up the fast reads in the queue, increasing the 99th percentile latencies significantly.

**Comment by nyh** [comment]
> The reader concurrency semaphore restricts the concurrency of reads that require CPU (intention: they read from the cache) to 1, meaning that if there is even a single active read which declares that it needs just CPU to proceed, no new read is admitted. This is meant to keep the concurrency of reads in the cache at 1. The idea is that concurrency in the cache is not useful: it just leads to the reactor rotating between these reads, all of the finishing later then they could if they were the only active read in the cache. This was observed to backfire in the case where there reads from a single table are mostly very fast, but on some keys are very slow (hint: collection full of tombstones). In this case the slow read keeps up the fast reads in the queue, increasing the 99th percentile latencies significantly.

Can you please remind me why we need **any** concurrency limit - be it 1 or 7 or 42 - on CPU-bound requests (i.e., read from cache)? The way I see it, if the request is very fast, it won't  yield at all and it doesn't matter what concurrency limit you set. But if the request is slow (e.g., needs to pass a thousand tombstones), and we yield (important to avoid stalls), then... so what? Who cares if we have 17 of these queries all having finished their CPU time quota and waiting for another one? Why is that bad, and how is better to do those 17 queries one after another?

If I understand correctly, the reason why we added a reader concurrency limit to reads that hold **disk buffers** is that we know that the disk has a limited concurrency - say, 10 reads concurrently, so it is pointless to have 100 requests all allocate disk buffers and then these 100 buffers will be held for a very long time - and it's better to only allocate 10 buffers at a time. But if you have a CPU-bound request which does not hold a disk buffer, this reason is no longer relevant.

**Comment by avikivity** [comment]
concurrency 1 with 100 x 10ms requests: latency 10ms, 20ms, 30ms, ... 1s. Requests that exceed their timeout can be aborted without consuming any resources.

concurrency 100 with 100 x 10ms requests: latency 1s, 1s, 1s, ... 1s. Requests that exceed their timeout can be aborted but already consumed resources.

**Comment by nyh** [comment]
> concurrency 1 with 100 x 10ms requests: latency 10ms, 20ms, 30ms, ... 1s. Requests that exceed their timeout can be aborted without consuming any resources.

Unless I'm missing something, this nice feature only happens on the first batch that comes while the machine is completely idle.. But Imagine that the 101st request comes while the first request is being processed - this 101st request will need to wait for all 100 before it to finish, i.e., 1s. In general, if the situation is that there's always 100 requests going on in parallel, the latency for every request will always be 1s.

Moreover, even in your special circumstance of 100 requests arriving together to an idle machine and then nothing more comes in until all are processed, the mean latency will still be 0.5s, it's not that much better than 1s.

> concurrency 100 with 100 x 10ms requests: latency 1s, 1s, 1s, ... 1s. Requests that exceed their timeout can be aborted but already consumed resources.

This last point is correct.

But Botond's point about the downside of the "concurrency 1" is also correct - it means that a particularly slow CPU-bound query (e.g., involving a lot of tombstones) will make all tiny queries wait. 
By the way, I don't know how increasing this 1 to 7 or 42 help. If you increase it to 42, then now if you have 42 concurrent slow queries will block out small ones. In other words, if you have a mix of slow and fast queries, the slow ones will start filling this queue while the fast ones finish quickly, until the point that the queue is filled with 42 slow queries.

**Comment by avikivity** [comment]
> > concurrency 1 with 100 x 10ms requests: latency 10ms, 20ms, 30ms, ... 1s. Requests that exceed their timeout can be aborted without consuming any resources.
> 
> Unless I'm missing something, this nice feature only happens on the first batch that comes while the machine is completely idle.. But Imagine that the 101st request comes while the first request is being processed - this 101st request will need to wait for all 100 before it to finish, i.e., 1s. In general, if the situation is that there's always 100 requests going on in parallel, the latency for every request will always be 1s.

It doesn't have to be completely idle. It's enough that a burst comes along. Even if initiation times are completely random, you will get bursts.

> 
> Moreover, even in your special circumstance of 100 requests arriving together to an idle machine and then nothing more comes in until all are processed, the mean latency will still be 0.5s, it's not that much better than 1s.

If the timeout is 0.5s, then with a concurrency of 1 50 requests will succeed and 50 will be rejected. With a concurrency of 100 all will fail.


> 
> > concurrency 100 with 100 x 10ms requests: latency 1s, 1s, 1s, ... 1s. Requests that exceed their timeout can be aborted but already consumed resources.
> 
> This last point is correct.
> 
> But Botond's point about the downside of the "concurrency 1" is also correct - it means that a particularly slow CPU-bound query (e.g., involving a lot of tombstones) will make all tiny queries wait. By the way, I don't know how increasing this 1 to 7 or 42 help. If you increase it to 42, then now if you have 42 concurrent slow queries will block out small ones. In other words, if you have a mix of slow and fast queries, the slow ones will start filling this queue while the fast ones finish quickly, until the point that the queue is filled with 42 slow queries.
